Userland/Libraries/LibGfx/ImageFormats/AVIFLoader.cpp

```diff
@@ -0,0 +1,206 @@
+/*
+ * Copyright (c) 2023, Nico Weber <thakis@chromium.org>
+ * Copyright (c) 2024, doctortheemh <doctortheemh@gmail.com>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <AK/Error.h>
+#include <LibGfx/ImageFormats/AVIFLoader.h>
+
+#include <avif/avif.h>
+
+namespace Gfx {
+
+struct AVIFLoadingContext {
+    enum State {
+        NotDecoded = 0,
+        Error,
+        HeaderDecoded,
+        BitmapDecoded,
+    };
+
+    State state { State::NotDecoded };
+    ReadonlyBytes data;
+    Optional<IntSize> size;
+    bool has_alpha { false };
+    size_t image_count { 0 };
+    size_t repetition_count { 0 };
+
+    Vector<ImageFrameDescriptor> frame_descriptors;
+};
+
+AVIFImageDecoderPlugin::AVIFImageDecoderPlugin(ReadonlyBytes data, OwnPtr<AVIFLoadingContext> context)
+    : m_context(move(context))
+{
+    m_context->data = data;
+}
+
+AVIFImageDecoderPlugin::~AVIFImageDecoderPlugin() = default;
+
+static ErrorOr<void> decode_avif_header(AVIFLoadingContext& context)
+{
+    if (context.state >= AVIFLoadingContext::HeaderDecoded)
+        return {};
+
+    avifDecoder* decoder = avifDecoderCreate();
+    if (decoder == NULL) {
+        return Error::from_string_literal("failed to allocate AVIF decoder");
+    }
+
+    avifResult result = avifDecoderSetIOMemory(decoder, context.data.data(), context.data.size());
+    if (result != AVIF_RESULT_OK) {
+        avifDecoderDestroy(decoder);
```

Comment:
If you're calling that for each exit path you can use a `ScopeGuard`:
```
ScopeGuard guard{[](){ avifDecoderDestroy(decoder); }};
```

---

As Zaggy suggested, it would be better to save that in the class and reuse the result in `decode_avif_image`

Reply:
Now reusing the decoder for all calls.